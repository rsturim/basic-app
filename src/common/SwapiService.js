/* eslint-disable compat/compat */

import axios from 'axios';

const baseUrl = 'https://swapi.co/api';

class SwapiService {
    getPlanets() {
        return axios.get(`${baseUrl}/planets/`);
    }
}
export default new SwapiService();
