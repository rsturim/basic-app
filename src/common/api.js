export default {
    addUser: (name, email, number) =>
        fetch(`www.simpsons.com`).then(res => res.json()),
};
