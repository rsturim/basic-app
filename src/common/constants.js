export const prefillContact = {
    name: 'John Doe',
    email: 'john@doe.com',
    number: '(802) 555-1212',
};

export const testBart = {
    name: 'Bart Simpson',
    email: 'bart@simpson.com',
    number: '(401) 245-3290',
};
