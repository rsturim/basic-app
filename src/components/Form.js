import React, { Component } from 'react';
import '../styles/form.scss';

import { prefillContact } from '../common/constants';

import api from '../common/api';

export default class Form extends Component {
    constructor(props) {
        super(props);

        this.defaultState = {
            name: '',
            email: '',
            number: '',
        };

        this.state = this.defaultState;
    }

    handleChange = str => e => {
        this.setState({ [str]: e.currentTarget.value });
    };

    handleSubmit = e => {
        e.preventDefault();
        return api.addUser(
            this.state.name,
            this.state.email,
            this.state.number,
        );
    };
    handlePromotionClick = e => {
        this.setState(prevState => ({ optIn: !prevState.optIn }));
    };

    prefill = e => {
        e.preventDefault();
        this.setState(prefillContact);
    };

    clearForm = e => {
        e.preventDefault();
        this.setState(this.defaultState);
    };

    render() {
        return (
            <form data-testid="addUserForm" onSubmit={this.handleSubmit}>
                <input
                    data-testid="name"
                    type="text"
                    placeholder="Name"
                    onChange={this.handleChange('name')}
                    value={this.state.name}
                />
                <input
                    data-testid="email"
                    type="text"
                    onChange={this.handleChange('email')}
                    placeholder="Email"
                    value={this.state.email}
                />
                <input
                    data-testid="number"
                    type="text"
                    onChange={this.handleChange('number')}
                    placeholder="Number"
                    value={this.state.number}
                />

                <div>
                    <button
                        type="submit"
                        className="btn btn-primary"
                        data-testid="submitButton"
                    >
                        Submit
                    </button>
                    <button
                        data-testid="clearButton"
                        className="btn btn-link"
                        onClick={this.clearForm}
                    >
                        Clear
                    </button>
                </div>

                <div>
                    <button
                        type="button"
                        className="btn btn-link"
                        data-testid="prefillButton"
                        onClick={this.prefill}
                    >
                        Pre-fill
                    </button>
                </div>
            </form>
        );
    }
}
