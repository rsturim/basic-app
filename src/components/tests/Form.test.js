import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { testBart as bart, prefillContact } from '../../common/constants';
import api from '../../common/api';
import Form from '../Form';

/* == form selectors */
// buttons
const submitButtonSelector = '[data-testid="submitButton"]';
const clearButtonSelector = '[data-testid="clearButton"]';
const prefillButtonSelector = '[data-testid="prefillButton"]';
// inputs
const nameFieldSelector = '[data-testid="name"]';
const emailFieldSelector = '[data-testid="email"]';
const numberFieldSelector = '[data-testid="number"]';

/* == constants */
const mockedEvent = { preventDefault: () => {} };

const updateInput = (wrapper, instance, newValue) => {
    const input = wrapper.find(instance);
    input.simulate('change', {
        currentTarget: { value: newValue },
    });
    return wrapper.find(instance);
};

describe('A form for signing up users', () => {
    test('should present buttons to interactive the form', () => {
        const wrapper = shallow(<Form />);

        expect(wrapper.find('button').length).toEqual(3);

        const submitButton = wrapper.find(submitButtonSelector);
        const clearButton = wrapper.find(clearButtonSelector);
        const prefillButton = wrapper.find(prefillButtonSelector);

        expect(submitButton.text()).toEqual('Submit');
        expect(clearButton.text()).toEqual('Clear');
        expect(prefillButton.text()).toEqual('Pre-fill');
    });

    test('should match the snapshot', () => {
        const shallowTree = shallow(<Form />);

        expect(toJson(shallowTree)).toMatchSnapshot();
    });

    describe('and when interacting with this form', () => {
        const wrapper = shallow(<Form />);
        const nameField = wrapper.find(nameFieldSelector);
        const emailField = wrapper.find(emailFieldSelector);
        const numberField = wrapper.find(nameFieldSelector);

        test('inputs to add you personal information should be available', () => {
            expect(wrapper.find('input').length).toEqual(3);

            expect(nameField.value).toEqual(undefined);
            expect(emailField.value).toEqual(undefined);
            expect(numberField.value).toEqual(undefined);
        });

        test('users can fill out form the form', () => {
            const wrapper = shallow(<Form />);
            const nameInput = updateInput(
                wrapper,
                nameFieldSelector,
                bart.name,
            );
            const emailInput = updateInput(
                wrapper,
                emailFieldSelector,
                bart.email,
            );
            const numberInput = updateInput(
                wrapper,
                numberFieldSelector,
                bart.number,
            );

            expect(nameInput.props().value).toEqual(bart.name);
            expect(emailInput.props().value).toEqual(bart.email);
            expect(numberInput.props().value).toEqual(bart.number);
        });

        test('users should be able to use the "pre-fill" button to populate the form', () => {
            const wrapper = shallow(<Form />);

            /* == test pre-fill */
            wrapper.find(prefillButtonSelector).simulate('click', mockedEvent);

            expect(wrapper.find(nameFieldSelector).props().value).toEqual(
                prefillContact.name,
            );
            expect(wrapper.find(emailFieldSelector).props().value).toEqual(
                prefillContact.email,
            );
            expect(wrapper.find(numberFieldSelector).props().value).toEqual(
                prefillContact.number,
            );

            /* == test clear */
            wrapper.find(clearButtonSelector).simulate('click', mockedEvent);

            expect(wrapper.find(nameFieldSelector).props().value).toEqual('');
            expect(wrapper.find(emailFieldSelector).props().value).toEqual('');
            expect(wrapper.find(numberFieldSelector).props().value).toEqual('');
        });

        //submits the form, calls api
        test('users can submit the form', () => {
            jest.spyOn(api, 'addUser').mockImplementation(() =>
                Promise.resolve({ data: 'New User Added' }),
            );
            const wrapper = shallow(<Form />);

            updateInput(wrapper, nameFieldSelector, bart.name);
            updateInput(wrapper, emailFieldSelector, bart.email);
            updateInput(wrapper, numberFieldSelector, bart.number);

            wrapper
                .find('[data-testid="addUserForm"]')
                .simulate('submit', { preventDefault: () => {} });

            expect(api.addUser).toHaveBeenCalledWith(
                bart.name,
                bart.email,
                bart.number,
            );
        });
    });
});
