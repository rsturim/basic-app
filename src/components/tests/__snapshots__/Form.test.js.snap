// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A form for signing up users should match the snapshot 1`] = `
<form
  data-testid="addUserForm"
  onSubmit={[Function]}
>
  <input
    data-testid="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value=""
  />
  <input
    data-testid="email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value=""
  />
  <input
    data-testid="number"
    onChange={[Function]}
    placeholder="Number"
    type="text"
    value=""
  />
  <div>
    <button
      className="btn btn-primary"
      data-testid="submitButton"
      type="submit"
    >
      Submit
    </button>
    <button
      className="btn btn-link"
      data-testid="clearButton"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
  <div>
    <button
      className="btn btn-link"
      data-testid="prefillButton"
      onClick={[Function]}
      type="button"
    >
      Pre-fill
    </button>
  </div>
</form>
`;
